zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,355 @@
+            msg = 'Stack name mismatch, {} != {}'.format(STACK_NAME,
+                                                         _stack.stack_name)
+            logging.error(msg)
+
+    def _stack_resource_compute(self):
+        """Confirm the stack has created a nova resource and check status."""
+        logging.info('Confirming heat stack resource status...')
+
+        # Confirm existence of a heat-generated nova compute resource.
+        _resource = self.heat_client.resources.get(STACK_NAME, RESOURCE_TYPE)
+        _server_id = _resource.physical_resource_id
+        if _server_id:
+            logging.debug('Heat template spawned nova instance, '
+                          'ID: {}'.format(_server_id))
+        else:
+            msg = 'Stack failed to spawn a nova compute resource (instance).'
+            logging.error(msg)
+
+        # Confirm nova instance reaches ACTIVE status.
+        openstack_utils.resource_reaches_status(self.nova_client.servers,
+                                                _server_id,
+                                                expected_status="ACTIVE",
+                                                msg="nova instance")
+        logging.info('Nova instance reached ACTIVE status')
+
+    def _stack_delete(self):
+        """Delete a heat stack, verify."""
+        logging.info('Deleting heat stack...')
+        openstack_utils.delete_resource(self.heat_client.stacks,
+                                        STACK_NAME, msg="heat stack")
+
+    def _image_delete(self):
+        """Delete that image."""
+        logging.info('Deleting glance image...')
+        image = self.nova_client.glance.find_image(IMAGE_NAME)
+        openstack_utils.delete_resource(self.glance_client.images,
+                                        image.id, msg="glance image")
+
+    def _keypair_delete(self):
+        """Delete that keypair."""
+        logging.info('Deleting keypair...')
+        openstack_utils.delete_resource(self.nova_client.keypairs,
+                                        KEYPAIR_NAME, msg="nova keypair")
+
+    def test_100_domain_setup(self):
+        """Run required action for a working Heat unit."""
+        # Action is REQUIRED to run for a functioning heat deployment
+        logging.info('Running domain-setup action on heat unit...')
+        unit = zaza.model.get_units(self.application_name)[0]
+        assert unit.workload_status == "active"
+        zaza.model.run_action(unit.entity_id, "domain-setup")
+        zaza.model.block_until_unit_wl_status(unit.entity_id, "active")
+        unit = zaza.model.get_unit_from_name(unit.entity_id)
+        assert unit.workload_status == "active"
+
+    def test_400_heat_resource_types_list(self):
+        """Check default resource list behavior and confirm functionality."""
+        logging.info('Checking default heat resource list...')
+        try:
+            types = list(self.heat_client.resource_types.list())
+            if type(types) is list:
```

Comment:
How could `type(types)` not be a `list`?  It was just set to one the line above?

Reply:
I suspect this is now wrong; `self.heat_client.resource_types.list()` may return a generator, not a list, which was why it was forced to a list.  You need to check the return type for that function and see what it can return to determine what the code below should be.